### DISCUSSION TITLE
006 test review

### DISCUSSION THREAD

**Discussion by cspayne** [uwlib-cams/uwlswd#52] [General Purpose]
@gerontakos @briesenberg07 
I'm putting the feedback from Theo's email, along with my own comments, here, so we can discuss before breaking things out into issues. 

0. RDF/XML
Faults in the RDF/XML
a. incorrect dct:hasFormat values. Removed by TG (in RDF/XML only).
b. Too bad order is not maintained. Old RDF/XML maintained the MARC standard's order, and that was helpful at times. I believe there's nothing we can do about that?
c. Comments were removed. Could they be maintained? The order is important for comments; they precede a sequence to which they pertain. The way they're written, I don't think we can keep them and process the RDF/XML, given that order is not preserved.
     
**CP: We can't maintain order with rdflib. If we want to maintain order (and comments), we could do the following:**
      **1. Make no changes to the rdf/xml during serialization - this means ensuring formatting is correct and in rdflib style _before_ serialization.**
      **2. Use a python xml processing library, such as lxml.etree to validate rdf/xml formatting before serialization.**
     **(comments and order would still not be maintained in other serializations)**

RDF/XML -->

1. HTML/RDFa
a. the html:

-Should not contain the triple dct:hasFormat=[itself] - **EDIT: fixed**
-values in Predicate column are not dereferencing properly
  dereference to url in prefix only.
    **CP: I found where this was happening. When run, these values will dereference properly now**

b. the RDFa
Superficially reviewed; looked ok.

c. the schema.org

- No language tag equivalents. Is that ok? - **I don't think schema.org has language tag equivalents**
- Only one creator, a URI, is present; did we decide on literal values here? - **no changes have been made yet**
- No alternateName present (i.e. alternative title) - **EDIT: added**
- Publisher: also only an IRI. I believe decision is pending on this?
- "@type" : "Dataset" --> I believe decision is pending on this?
- contributor property spelled contributer. - **spelling fixed now**
- Only contributors with IRIs are present. Incorrect? - **currently, yes**
- No isBasedOn (i.e. dct:source) present. - **EDIT: added**
- In 006/form_of_material.html, "description" : "VALUE MISSING" seems incorrect - **should be working now**

3. JSON-LD

    When RDF is indented, new lines show up in JSON-LD (i.e. \n) - 
        solution: RDF/XML should not be indented.

**CP: This is interesting. I haven't come across this in previous tests, so I'm not sure what's happening here.**

4. N-Triples (same indentation problem; otherwise OK)

5. Turtle (same indentation problem)

6. Datacite XML
- Value of resourceType is not correct (should be some phrase; I don't think we've resolved how to do this.)
- //title/@titleType for alt title should be AlternativeTitle not Alternative Title. - **fixed**
- //relatedIdentifiers/relatedIndentifier should be //relatedIdentifiers/relatedIdentifiers  - **fixed**

- //relatedIdentifier/@relationType="isDerivedFrom" should be //relatedIdentifier/@relationType="IsDerivedFrom" - **fixed capitalization**

**Comment by briesenberg07** [comment]
Thank you @cspayne ! Responding with thoughts, yea/nay votes, asides, etc. below. Not responding to things that are settled, or settled to be fixed, etc.

☝🏻 1️⃣ **Maintaining triples order + comments in RDF/XML**

> b. Too bad order is not maintained. Old RDF/XML maintained the MARC standard's order, and that was helpful at times. I believe there's nothing we can do about that?
> c. Comments were removed. Could they be maintained? The order is important for comments; they precede a sequence to which they pertain. The way they're written, I don't think we can keep them and process the RDF/XML, given that order is not preserved.

*I understand CP to say that if we want to preserve order of triples, and/or preserve XML comments, we'd need to process RDF/XML as XML and not RDF. I vote **nay** for doing this, because I have much less confidence in processing XML with Python than I do with processing RDF. BUT I don't fully understand the needs here: I think I understand the argument for preserving order--this would present the concepts in the same order as they are presented elsewhere as MARC codes and make the data more human-readable?--but I don't understand the argument for comments--what is the content of comments?*

☝🏻 2️⃣ **schema.org stuff**

> No language tag equivalents. Is that ok? - I don't think schema.org has language tag equivalents

*Wow way to be Anglocentric schema.org. No language tags!?*

> Only one creator, a URI, is present; did we decide on literal values here? - no changes have been made yet
> Only contributors with IRIs are present. Incorrect? - currently, yes

#49 *relates to above and to what we'll output for schema.org markup - I need to follow up on @gerontakos ' comments there - but I think this decision should be recorded in [rdfxml-datacite-schema.org](https://docs.google.com/spreadsheets/d/1jKYPFzhE_iNsqrekX696_wFehq9NTM4v212J41jGv0s/edit?usp=sharing)*

> "@type" : "Dataset" --> I believe decision is pending on this?

*Need to record in [rdfxml-datacite-schema.org](https://docs.google.com/spreadsheets/d/1jKYPFzhE_iNsqrekX696_wFehq9NTM4v212J41jGv0s/edit?usp=sharing), I'm 99% sure above is correct but still need to confirm. I'm thrown off because I don't see any schema:type property, I think schema uses rdf:type??*

☝🏻 3️⃣ **Don't indent RDF/XML** (!? 😢 )

> When RDF is indented, new lines show up in JSON-LD (i.e. \n) - solution: RDF/XML should not be indented.
> N-Triples (same indentation problem; otherwise OK)
> Turtle (same indentation problem)

*Is this a problem with the indentation in the 'fancy HTML' preview pane for serializations? Hadn't noticed any problems. I need to take a look...*

☝🏻 4️⃣ **datacite:resourceType**

> Value of resourceType is not correct (should be some phrase; I don't think we've resolved how to do this.)

*This is on the agenda - BMR [proposes including BMPs in RDF/XML using schema:disambiguating description](https://docs.google.com/document/d/13OK-OqBnqaihnfByu-wZa65F0AlpVTrUYHWZs2SZAUk/edit#bookmark=id.plf4u1kgxjcz)*

**Comment by cspayne** [comment]
> JSON-LD
When RDF is indented, new lines show up in JSON-LD (i.e. \n) -
solution: RDF/XML should not be indented.

@gerontakos are you referring to instances within the literal object of triples? Such as:
```json 
    "http://www.w3.org/2004/02/skos/core#definition": [
      {
        "@language": "en",
        "@value": "Three-dimensional artifact or naturally occurring\n         object."
      }
```

In this case, the serialization is maintaining a new line element that was [already in the original rdf/xml](https://github.com/uwlib-cams/uwlswd_vocabs_marc_006_008/blob/989bd0084a237865f71feea6e76c61780c703be0/006/form_of_material.rdf#L147C7-L147C7):
```
      <skos:definition xml:lang="en">Three-dimensional artifact or naturally occurring
         object.</skos:definition>
```

I think as long as there aren't new line elements within the literal values, we should still be able to indent rdf/xml without any issue.

**Comment by gerontakos** [comment]
1️⃣ Maintaining triples order + comments in RDF/XML
I almost always argue against "order" in RDF; I would like to maintain that here. We should not construct RDF data where the order matters! That is not RDF. That's more an XML thing. As for comments: we have no need for comments, and going forward, we should not use comments to explain anything in the RDF/XML. In the 008 vocabs, the comments are not essential. I was thinking only that David included those comments to make the vocabularies a little easier to look at, and it would be affirmation of his judgment if we included them. But let's forget about order and comments in RDF/XML going forward.
 2️⃣ schema.org stuff
I thought we could create an object for each literal value.
```
"creator" : {
    "@value" : "Metadata Management Associates" ;
    "@language" : "en "
}
```
 3️⃣ Don't indent RDF/XML
Indenting causes line breaks -- but it must vary, depending on the XML editor and also on the settings within each editor. I use oXygen, and it consistently indents (with my current settings, which are default settings I'm not aware of) the following:
```
<skos:scopeNote>Use for newspapers created by schools, but generally not by the student body, e.g., Indian school newspapers.</skos:scopeNote>> 
```
as
``` 
<skos:scopeNote>Use for newspapers created by schools, but generally not by the student
            body, e.g., Indian school newspapers.</skos:scopeNote>
```
producing json-ld like:
```
"@value": "Use for newspapers created by schools, but generally not by the student\n      body, e.g. Indian school newspapers"
```
I think generally our lines of RDF/XML are not long enough to cause the indentation to insert white space on indentation; however, the 006 vocabulary had a few admittedly not-so-long lines that inserted white space on indent.

Oh I just saw Cypress' latest post; yes, that's the problem, and presumably we can watch for "broken" strings of text in the RDF/XML, I'm not sure how. I do know the normalize-space() function in XPath eliminates consecutive whitespace characters in a literal value.  

I don't think the loss is tremendous if the RDF/XML is not indented. It's difficult to view in Github; well, you can just clone and view in your favorite editor, right?

Finally:
4:
I expect you realized I made an error on my datacite bullet 3 (I thought I'd fixed that):
//relatedIdentifiers/relatedI**n**dentifier should be //relatedIdentifiers/relatedIdentifiers
...should have been:
//relatedIdentifiers/relatedI**n**dentifier should be //relatedIdentifiers/relatedIdentifier

**Comment by cspayne** [comment]
>  I use oXygen, and it consistently indents (with my current settings, which are default settings I'm not aware of)

In oXygen, If you go to options > preferences > editor > format > xml and check the option for 'preserve text as it is', it will prevent indentation in the middle of literals!

**Reply by gerontakos** [reply]
Oh wau, it works!

**Comment by cspayne** [comment]
re: schema.org
Looking at the [Thing type](https://schema.org/Thing) (of which Person and Creator are subtypes of), I'm still not certain that we can do something like:
```
"creator" : {
    "@value" : "Metadata Management Associates" ;
    "@language" : "en "
}
```
But if it is permissible, then I am all for it!

**Reply by gerontakos** [reply]
Oh I made a double-error. First, I was thinking JSON-LD; second, the creator themself is not in English! (Only the name is.) I retract the proposal!

**Comment by briesenberg07** [comment]
This was initial feedback on the main.py workflow, closing this discussion. For further feedback on workflow create new issue/discussion
